lib/query.js

```diff
@@ -72,4 +116,112 @@ export const getDenominator = (node) => {
+    if (isNumber(node)) {
+        return node
+    } else if (isIdentifier(node) || isPow(node)) {
+        return build.numberNode('1')
+    } else if (isNeg(node)) {
+        const result = build.applyNode('neg', [getCoefficient(node.args[0])])
+        result.wasMinus = node.wasMinus
+        return result
+    } else if (isMul(node)) {
+        const numbers = node.args.filter(arg => isNumber(arg) || isConstantFraction(arg))
+        if (numbers.length > 1) {
+            return build.mul(...numbers)
+        } else if (numbers.length > 0) {
+            return numbers[0]
+        } else {
+            return build.numberNode('1')
+        }
+    }
+}
+
+export const getVariableFactors = (node) => {
+    if (isVariableFactor(node)) {
+        return [node]
+    } else if (isMul(node)) {
+        return node.args.filter(isVariableFactor)
+    } else if (isNeg(node)) {
+        // TODO: figure out how to handle (x)(-y)(z)
+        return getVariableFactors(node.args[0])
+    } else {
+        return []
+    }
+}
+
+export const getVariableFactorName = (node) => {
+    if (isIdentifier(node)) {
+        return node.name
+    } else if (isPow(node)) {
+        return getVariableFactorName(node.args[0])
+    } else {
+        return null
+    }
+}
+
+export const sortVariables = (variables) =>
+    variables.sort(
+        (a, b) => getVariableFactorName(a) > getVariableFactorName(b))
+
+export const hasConstantBase = node => isPow(node) && isNumber(node.args[0])
+
+export const getCoefficientsAndConstants = (node, coefficientMap = {}, constants = [], others = []) => {
+    if (isNumber(node) || isConstantFraction(node)) {
+        constants.push(node)
+    } else if(isPolynomialTerm(node) && !hasConstantBase(node)) {
+        const sortedVariables = sortVariables(getVariableFactors(node))
+
+        const coefficient = getCoefficient(node)
+        const implicit = isImplicit(node)
+        const key = sortedVariables.length > 1
+              ? print(build.applyNode('mul', sortedVariables, {implicit}))
+              : print(sortedVariables[0])
+
+        if (!(key in coefficientMap)) {
+            coefficientMap[key] = [coefficient]
+        } else {
+            coefficientMap[key].push(coefficient)
+        }
+    } else if(isPolynomial(node) || isApply(node)) {
+        node.args.forEach(function(arg) {
+            getCoefficientsAndConstants(arg, coefficientMap, constants, others)
+        })
+    } else {
+        others.push(node)
+    }
+
+    return {coefficientMap, constants, others}
+}
+
+// returns true if two nodes have the same base variable
+// e.g. (x + 1)^1 and (x + 1)^3 , true
+// e.g. x^2, (x + 1)^2, false
+export const hasSameBase = (node1, node2) => {
+    return print(node1) == print(node2)
+        || isPow(node1) && print(node1.args[0]) == print(node2)
+        || isPow(node2) && print(node1) == print(node2.args[0])
+        || isPow(node1) && isPow(node2) && print(node1.args[0]) == print(node2.args[0])
```

Comment:
Now that you've added `nodeEquals` this can be:
```
return nodeEquals(node1, node2) 
    || isPow(node1) && nodeEquals(node1.args[0], node)
    || isPow(node2) && nodeEquals(node1, node2.args[0])
    || isPow(node1) && isPow(node2) && nodeEquals(node1.args[0], node2.args[0])
```